### PR TITLE
Limit the number of errors output in a single run to 256

### LINKED ISSUE #2876: Limit number of errors reported to something smallish (100).
Sometimes a small typo results in pages and pages of errors in Solidity compiler output.  The compiler should limit the number of errors shown to a smallish number (maybe 100), as additional errors after that are not useful and end up blowing stderr scroll buffer making it so relevant information may no longer be available.

### PR DESCRIPTION
**Pull request by axic**
Fixes #2876

### CODE REVIEW

File: libsolidity/interface/ErrorReporter.cpp [hunk 1]
```diff
@@ -80,6 +84,16 @@ void ErrorReporter::error(Error::Type _type, SourceLocation const& _location, Se
 	m_errorList.push_back(err);
 }
 
+void ErrorReporter::abortIfExcessive()
+{
+	if (m_errorList.size() > 256)
+	{
+		auto err = make_shared<Error>(Error::Type::Warning);
+		*err << errinfo_comment("Dude, there are way too many errors. Chill out.");
```
Comment (federicobond) [inline libsolidity/interface/ErrorReporter.cpp]: I have nothing against having some fun with the messages, but I would prefer a less gendered term than "Dude".

**Comment by federicobond** [comment]
It would be cleaner to handle this at output time, when you know how many errors were reported and can say things like: "X errors ommited."

**Comment by federicobond** [comment]
Some clients may also want to exercise precise control over error output.

**Comment by axic** [comment]
> It would be cleaner to handle this at output time, when you know how many errors were reported and can say things like: "X errors ommited."

Wasn't the main point of this report the excessive time it takes to create the errors as opposed to simply writing it out?

In some of the original reports it may have taken up to 2-3GB of memory to keep all the errors :)

**Comment by federicobond** [comment]
Yes, technically you could keep compiling without storing the messages, but bailing out looks like the reasonable choice here.

File: libsolidity/interface/ErrorReporter.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/interface/ErrorReporter.cpp]: We always have to be super formal 😭

File: libsolidity/interface/ErrorReporter.cpp [hunk 2]
```diff
@@ -80,6 +84,16 @@ void ErrorReporter::error(Error::Type _type, SourceLocation const& _location, Se
...
 }
 
+void ErrorReporter::abortIfExcessive()
+{
+	if (m_errorList.size() > 256)
+	{
+		auto err = make_shared<Error>(Error::Type::Warning);
+		*err << errinfo_comment("There are more than 256 errors. Aborting.");
+		m_errorList.push_back(err);
+		BOOST_THROW_EXCEPTION(FatalError());
```
Comment (chriseth) [inline libsolidity/interface/ErrorReporter.cpp]: In this case, we need to add a fatal error catcher into every sub-component, like SyntaxChecker.

File: libsolidity/interface/ErrorReporter.cpp [hunk 3]
```diff
@@ -80,6 +84,16 @@ void ErrorReporter::error(Error::Type _type, SourceLocation const& _location, Se
 	m_errorList.push_back(err);
 }
 
+void ErrorReporter::abortIfExcessive()
+{
+	if (m_errorList.size() > 256)
+	{
+		auto err = make_shared<Error>(Error::Type::Warning);
```
Comment (chriseth) [inline libsolidity/interface/ErrorReporter.cpp]: This cannot be a warning, because it might be that we have 256 warnings and errors appear only later.

Comment (axic) [reply libsolidity/interface/ErrorReporter.cpp]: Not happy with the type, but what is the best type for it?

Comment (axic) [reply libsolidity/interface/ErrorReporter.cpp]: Actually I was wondering if this should exclude (ignore) warnings when calculating the limit.

Comment (axic) [reply libsolidity/interface/ErrorReporter.cpp]: It seems that (ignoring warnings) is the way to go:

```
,truffle/Assert.sol:640:43: Warning: Function declared as view, but this expression (potentially) modifies the state and thus requires non-payable (the default) or payable.
            _report(result, _appendTagged(_tag(A, "Tested"), _tag(B, "Against"), message));
                                          ^---------------^
,Warning: There are more than 256 errors. Aborting.

Error: Could not find artifacts for ./TestMath.sol from any sources
    at Resolver.require (/tmp/tmp.H6c4gxXTth/node_modules/truffle/build/cli.bundled.js:50099:9)
    at TestResolver.require (/tmp/tmp.H6c4gxXTth/node_modules/truffle/build/cli.bundled.js:96057:30)
    at /tmp/tmp.H6c4gxXTth/node_modules/truffle/build/cli.bundled.js:209768:30
    at Array.map (native)
    at /tmp/tmp.H6c4gxXTth/node_modules/truffle/build/cli.bundled.js:209766:33

  Command failed: truffle test --network testing
```

File: libsolidity/interface/ErrorReporter.cpp [hunk 4]
```diff
@@ -80,6 +86,37 @@ void ErrorReporter::error(Error::Type _type, SourceLocation const& _location, Se
...
 
+bool ErrorReporter::checkForExcessiveErrors(Error::Type _type)
+{
+	if (_type == Error::Type::Warning)
+	{
+		m_warningCount++;
+
+		if (m_warningCount == c_maxWarningsAllowed)
+		{
+			auto err = make_shared<Error>(Error::Type::Warning);
```
Comment (chriseth) [inline libsolidity/interface/ErrorReporter.cpp]: But now as we split into warnings and errors, this should be a warning. I can fix that.

Comment (chriseth) [reply libsolidity/interface/ErrorReporter.cpp]: Ah forget it, it _is_ a warning.

**Comment by axic** [comment]
Oh this was failing a test, but I think that was due to the a momentary RPC issue.

Also, the commit `reindent` should have been squashed to the preceding commit, just left it separate for review purposes :)

**Pull request merged**